### PR TITLE
Solution for week 4 # hoangdangduy

### PR DESCRIPTION
**Pull request by hoangdangduy**
Awesome when I do according to your video guide to create new branch and success :D

### CODE REVIEW

File: week0004/hoangdangduy/java/Week0004.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+				varRun++;
+			}else{
+				if (varRun != 0){
+					sortUtil.QuickSoft(listClient, i-varRun, i, "email");
+				}
+				varRun = 0;
+			}
+		}
+		
+		for (int i=0; i<listClient.size(); i++){
```
Comment (vietlq) [inline week0004/hoangdangduy/java/Week0004.java]: Spacing

File: week0004/hoangdangduy/java/SortUtil.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+				j--;
+			}
+		} while (i <= j);
+		t = listClient.get(left);
+		listClient.set(left, listClient.get(j));
+		listClient.set(j, t);
+		return j;
+	}
+
+	void QuickSoft(ArrayList<Client> listClient, int left, int right, String chooseFunctionSort) {
```
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: Quick sort?

File: week0004/hoangdangduy/java/SortUtil.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import java.util.ArrayList;
+
+public class SortUtil {
+	public int Partition(ArrayList<Client> listClient, int left, int right, String chooseFunctionSort) {
```
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: This looks complicated. Think about comparators

File: week0004/hoangdangduy/java/Week0004.java (hunk 1 above)
Comment (hoangdangduy) [inline week0004/hoangdangduy/java/Week0004.java]: I think it will make it easily and clearly when read code. Is it not good in real project?

**Comment by vietlq** [comment]
Always have spaces around operators, between ( {

File: week0004/hoangdangduy/java/Week0004.java (hunk 1 above)
Comment (vietlq) [inline week0004/hoangdangduy/java/Week0004.java]: ) {

Comment (vietlq) [inline week0004/hoangdangduy/java/Week0004.java]: Spaces around operators

Comment (vietlq) [inline week0004/hoangdangduy/java/Week0004.java]: You should read professional code and see code styles. Spacing is important.

File: week0004/hoangdangduy/java/SortUtil.java (hunk 2 above)
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: You can use comparator to achieve the same effect and keep your code clean.

Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html

https://docs.oracle.com/javase/7/docs/api/java/util/Comparator.html

static <T> void	sort(T[] a, Comparator<? super T> c)
Sorts the specified array of objects according to the order induced by the specified comparator.

File: week0004/hoangdangduy/java/SortUtil.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Comparator;
+
+public class SortUtil {
+
+	public ArrayList<Client> sort(ArrayList<Client> listClient) {
+		Collections.sort(listClient, new Comparator<Client>() {
+
+			@Override
+			public int compare(Client client1, Client client2) {
+				return client1.getFname().compareTo(client2.getFname());
```
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: Compare last name if first name is the same. Compare email if last name is the same

File: week0004/hoangdangduy/java/SortUtil.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			}
+		});
+
+		int varRun = 0;
+		/*
+		 * Check fname duplicate to sort lname according to fname
```
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: Remove this part. See above

File: week0004/hoangdangduy/java/SortUtil.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+							return client1.getLname().compareTo(client2.getLname());
+						}
+					});
+				}
+				varRun = 0;
+			}
+		}
+
+		/*
+		 * Check lname duplicate to sort email according to lname
```
Comment (vietlq) [inline week0004/hoangdangduy/java/SortUtil.java]: Remove this part. See above

File: week0004/hoangdangduy/java/SortUtil.java (hunk 4 above)
Comment (hoangdangduy) [inline week0004/hoangdangduy/java/SortUtil.java]: Yes. Do my code process wrong?. First, I sort client order by fname then I sort lname when have similar fname and email do same before.

**Comment by vietlq** [comment]
Awesome!!!! Hope you enjoyed 4 weeks of coding! Hope to see you in the upcoming class!

**Pull request merged**